[REVIEW] release HRP greenflow plugin

yidong72 opened the pull request:
<!--

Thank you for contributing to gQuant :)

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. There are CI checks in place to enforce that committed code follows our style
   and syntax standards. Please see our contribution guide in `CONTRIBUTING.MD`
   in the project root for more information about the checks we perform and how
   you can run them locally.

4. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

5. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]`.

6. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove the `[WIP]` label (if present) and replace
   it with `[REVIEW]`. If assistance is required to complete the functionality,
   for example when the C/C++ code of a feature is complete but Python bindings
   are still required, then add the label `[HELP-REQ]` so that others can triage
   and assist. The additional changes then can be implemented on top of the
   same PR. If the assistance is done by members of the rapidsAI team, then no
   additional actions are required by the creator of the original PR for this,
   otherwise the original author of the PR needs to give permission to the
   person(s) assisting to commit to their personal fork of the project. If that
   doesn't happen then a new PR based on the code of the original PR can be
   opened by the person assisting, which then will be the PR that will be
   merged.

7. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please do not
   rebase your branch on main/force push/rewrite history, doing any of these
   causes the context of any comments made by reviewers to be lost. If
   conflicts occur against main they should be resolved by merging main
   into the branch used for making the pull request.

Many thanks in advance for your cooperation!

-->
Release the code for HRP project.

avolkov1 reviewed (changes requested):
One of the unit tests failed.
```
test_max_drawdown (unit.test_max_drawdown.TestMaxDrawdown) ... FAIL
```

Minor flake warnings.
```
$ flake8 greenflow_hrp_plugin/
greenflow_hrp_plugin/__init__.py:41:79: W291 trailing whitespace
greenflow_hrp_plugin/kernels.py:592:80: E501 line too long (81 > 79 characters)
greenflow_hrp_plugin/loadCsvNode.py:109:13: F841 local variable 'client' is assigned to but never used
```

I'm not done reviewing. Some comments are just remarks. So far the unit test above should be fixed.

yidong72 commented on gQuant/plugins/hrp_plugin/docker/Dockerfile:
```diff
@@ -0,0 +1,28 @@
+ENV PATH="/root/miniconda3/bin:${PATH}"
+RUN conda init
+RUN conda install -y -c conda-forge python-graphviz nodejs=12.4.0  pydot
+RUN conda install -y -c rapidsai -c nvidia -c conda-forge -c defaults rapids=0.19
```
Yes. I will upgrade it to the latest version.

yidong72 commented on gQuant/plugins/hrp_plugin/README.md:
````diff
@@ -0,0 +1,99 @@
+
+#### Install the latest RAPIDS
+```bash
+conda install -y -c rapidsai -c nvidia -c conda-forge -c defaults rapids=0.17
````
agree.

yidong72 commented on gQuant/plugins/hrp_plugin/greenflow_hrp_plugin/featureNode.py:
```diff
@@ -0,0 +1,122 @@
+            col.remove('year')
+            col.remove('month')
+            mean_val = df[col].values.mean(axis=1)
+            std_val = df[col].values.mean(axis=1)
```
good catch!

yidong72 commented on gQuant/plugins/hrp_plugin/greenflow_hrp_plugin/aggregateTimeFeature.py:
```diff
@@ -0,0 +1,135 @@
+        df = inputs[self.INPUT_PORT_NAME]
+        # df = df.drop('datetime', axis=1)
+        output = {}
+        if self.outport_connected(self.OUTPUT_PORT_NAME):
```
Yes. it is not necessary. I will remove the check

yidong72 commented on gQuant/plugins/hrp_plugin/greenflow_hrp_plugin/transactionCostNode.py:
```diff
@@ -0,0 +1,113 @@
+        cost_df['month'] = df['month']
+        cost_df['sample_id'] = df['sample_id']
+        output = {}
+        if self.outport_connected(self.OUTPUT_PORT_NAME):
```
I will remove the if check

yidong72 commented:
addressed all the comments. RAPIDS 21.08 will break the dask_cudf dataframe type. RAPIDS 21.06 is fine.
Need to fix the `from_delayed` method to generate proper type in the future.

avolkov1 reviewed (commented):
There's a slight nuisance in the notebooks. Some of the workflows produce a warning that port is missing in incoming meta data:
```
Task "max_drawdown_feature" Node Type "[<class 'greenflow_hrp_plugin.featureNode.FeatureNode'>]" missing required port "feature_df" in incoming meta data. Should the port be connected?
Task "all_max_drawdown_feature" Node Type "[<class 'greenflow_hrp_plugin.featureNode.FeatureNode'>]" missing required port "feature_df" in incoming meta data. Should the port be connected?
```

The reason is that the node `greenflow_hrp_plugin.featureNode.FeatureNode` has a port `feature_df` that is optional, i.e. it is not required to be connected (but if it is connected its data is used).
For such a scenario there is a port spec called "optional" (`PortsSpecSchema.optional: True`). I tweaked the code to get this working to inhibit the warning. These are the changes I made:
```
diff --git a/gQuant/plugins/hrp_plugin/greenflow_hrp_plugin/featureNode.py b/gQuant/plugins/hrp_plugin/greenflow_hrp_plugin/featureNode.py
index 8a8da3f..ba6b6f1 100644
--- a/gQuant/plugins/hrp_plugin/greenflow_hrp_plugin/featureNode.py
+++ b/gQuant/plugins/hrp_plugin/greenflow_hrp_plugin/featureNode.py
@@ -40,7 +40,8 @@ class FeatureNode(TemplateNodeMixin, Node):
                 port_type: [
                     "pandas.DataFrame", "cudf.DataFrame",
                     "dask_cudf.DataFrame", "dask.dataframe.DataFrame"
-                ]
+                ],
+                PortsSpecSchema.optional: True
             }
         }
         port_outports = {
diff --git a/greenflow/greenflow/dataframe_flow/_node_flow.py b/greenflow/greenflow/dataframe_flow/_node_flow.py
index 6fda95f..749c64f 100644
--- a/greenflow/greenflow/dataframe_flow/_node_flow.py
+++ b/greenflow/greenflow/dataframe_flow/_node_flow.py
@@ -858,6 +858,10 @@ class NodeTaskGraphMixin(NodeTaskGraphExtensionMixin):
                 dy = PortsSpecSchema.dynamic
                 if inports[iport].get(dy, False):
                     continue
+
+                if inports[iport].get(PortsSpecSchema.optional, False):
+                    continue
+
                 # Is it possible that iport not connected? If so iport should
                 # not be in required. Should raise an exception here.
                 warn_msg = \
diff --git a/greenflow/greenflow/dataframe_flow/_node_taskgraph_extension_mixin.py b/greenflow/greenflow/dataframe_flow/_node_taskgraph_extension_mixin.py
index 0363df0..9be1fc4 100644
--- a/greenflow/greenflow/dataframe_flow/_node_taskgraph_extension_mixin.py
+++ b/greenflow/greenflow/dataframe_flow/_node_taskgraph_extension_mixin.py
@@ -1,3 +1,4 @@
+from copy import deepcopy
 from .portsSpecSchema import (PortsSpecSchema, NodePorts)
 from .metaSpec import (MetaDataSchema, MetaData)
 
@@ -56,12 +57,10 @@ class NodeTaskGraphExtensionMixin:
         dynamic = None
         inports = {}
         for input_port in port_inports:
+            inports[input_port] = deepcopy(port_inports[input_port])
             if input_port in input_connections:
                 determined_type = input_connections[input_port]
-                inports[input_port] = {port_type: determined_type}
-            else:
-                types = port_inports[input_port][port_type]
-                inports[input_port] = {port_type: types}
+                inports[input_port].update({port_type: determined_type})
 
             if dy in port_inports[input_port]:
                 inports[input_port][dy] = True
```

I set that port as optional in `featureNode.py`. Then I had to tweak the logic a bit in `_node_flow.py:_validate_connected_metadata`, and `_node_taskgraph_extension_mixin.py:ports_setup_ext` to account for the `PortsSpecSchema.optional` case.

I did not test this thoroughly, but the notebooks seem to run fine and the warning is now correctly handled. 

What do you think of making these updates?

avolkov1 reviewed (changes requested):
I finished reviewing. There's a bug when drawing the taskgraph with ports in ipynb mode. 
```
taskGraph.draw(show='ipynb', show_ports=True)
```
That draw method is not called with `show_ports` option in the notebooks, nevertheless it can be fixed via:
```
diff --git a/greenflow/greenflow/dataframe_flow/taskGraph.py b/greenflow/greenflow/dataframe_flow/taskGraph.py
index 70149b1..66fc0ff 100644
--- a/greenflow/greenflow/dataframe_flow/taskGraph.py
+++ b/greenflow/greenflow/dataframe_flow/taskGraph.py
@@ -353,7 +353,7 @@ class TaskGraph(object):
 
                 if (to_type == OUTPUT_TYPE):
                     continue
-                task_node = get_node_obj(itask)
+                task_node = get_node_obj(itask, tgraph_mixin=True)
```

I had trouble running through the 17assets.ipynb notebook. Even reducing the bootstrap samples I would run out of memory on the GPU. I had to re-start the notebook kernel and run each section individually, and that way I was able to run through the whole notebook. I think there's a bug with `cudf.merge` as that method seems to cause memory leak on the GPUs.

My suggestion would be to split that notebook "17assets.ipynb" into several notebooks corresponding to 1 or 2 sections at a time. Name them something like "17assets_1of5.ipynb", "17assets_2of5.ipynb", and so on, or whatever makes sense. That might make it easier to run each notebook, clean the python kernel or shutdown the notebook, run the next notebook, and so on.

Let me know if you'd like to make these changes or not. I'll resolve this comment then, and approve/merge the PR.

Thanks

yidong72 commented:
> I finished reviewing. There's a bug when drawing the taskgraph with ports in ipynb mode.
> 
> ```
> taskGraph.draw(show='ipynb', show_ports=True)
> ```
> 
> That draw method is not called with `show_ports` option in the notebooks, nevertheless it can be fixed via:
> 
> ```
> diff --git a/greenflow/greenflow/dataframe_flow/taskGraph.py b/greenflow/greenflow/dataframe_flow/taskGraph.py
> index 70149b1..66fc0ff 100644
> --- a/greenflow/greenflow/dataframe_flow/taskGraph.py
> +++ b/greenflow/greenflow/dataframe_flow/taskGraph.py
> @@ -353,7 +353,7 @@ class TaskGraph(object):
>  
>                  if (to_type == OUTPUT_TYPE):
>                      continue
> -                task_node = get_node_obj(itask)
> +                task_node = get_node_obj(itask, tgraph_mixin=True)
> ```
> 
> I had trouble running through the 17assets.ipynb notebook. Even reducing the bootstrap samples I would run out of memory on the GPU. I had to re-start the notebook kernel and run each section individually, and that way I was able to run through the whole notebook. I think there's a bug with `cudf.merge` as that method seems to cause memory leak on the GPUs.
> 
> My suggestion would be to split that notebook "17assets.ipynb" into several notebooks corresponding to 1 or 2 sections at a time. Name them something like "17assets_1of5.ipynb", "17assets_2of5.ipynb", and so on, or whatever makes sense. That might make it easier to run each notebook, clean the python kernel or shutdown the notebook, run the next notebook, and so on.
> 
> Let me know if you'd like to make these changes or not. I'll resolve this comment then, and approve/merge the PR.
> 
> Thanks
I have done all the changes.  I added following lines to restart the kernel which seems to fix the memory issue.
```python
import IPython
app = IPython.Application.instance()
app.kernel.do_shutdown(True)
```
Could you give it a try?

yidong72 commented:
ready for review

avolkov1 reviewed (commented):
Thank you. This looks good. I was able to run the notebook. I had to be a little patient after running the kernel shutdown cell. Otherwise this worked.

Did you re-run the 17assets.ipynb notebook completely? Because in the gtihub posted version (https://github.com/yidong72/gQuant/blob/branch-hrp/gQuant/plugins/hrp_plugin/notebooks/17assets.ipynb) the sections after "Hyperparameter Optimization" have the warning that was fixed. Running it myself, and none of those warnings appeared anymore.

yidong72 commented:
I added your 17assets notebook and added the synthetic 10 stocks dataset.

avolkov1 reviewed (approved):
Looks great. Thanks!
